plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,231 @@
+   }
+
+   bool is_any_filesystem_space_below_threshold() {
+      // Go over each monitored file system
+      for (auto& fs: filesystems) {
+         boost::system::error_code ec;
+         auto info = bfs::space(fs.path_name, ec);
+         if ( ec ) {
+            // As the system is running and this plugin is not a critical
+            // part of the system, we should not exit.
+            // Just report the failure and continue;
+            wlog( "Unable to get space info for ${path_name}: [code: ${ec}] ${message}",
+               ("path_name", fs.path_name.string())
+               ("ec", ec.value())
+               ("message", ec.message()));
+
+            continue;
+         }
+
+         if (info.available < fs.min_available_bytes) {
+            wlog("Space usage on ${path}'s file system exceeded configured threshold ${threshold}%, initiate shuting down!!! available: ${available}, Capacity: ${capacity}, min_available_bytes: ${min_available_bytes}", ("path", fs.path_name.string()) ("threshold", used_space_threshold_in_percentage) ("available", info.available) ("capacity", info.capacity) ("min_available_bytes", fs.min_available_bytes));
+
+            return true;
+         }
+      }
+
+      return false;
+   }
+
+   void add_file_system(const bfs::path& path_name) {
+      // Get detailed information of the path
+      struct stat statbuf;
+      auto status = stat(path_name.string().c_str(), &statbuf);
+      EOS_ASSERT(status == 0, chain::plugin_config_exception,
+                 "Failed to run stat on ${path} with status ${status}", ("path", path_name.string())("status", status));
+
+      // If the file system containing the path is already
+      // in the filesystem list, do not add it to the list again
+      for (auto& fs: filesystems) {
+         if (statbuf.st_dev == fs.st_dev) { // Two files belong to the same file system if their device IDs are the same.
+            ilog("file ${path_name}'s file system already in list", ("path_name", path_name.string()));
+
+            return;
+         }
+      }
+
+      // For efficeincy, precalculate threshold values to avoid calculating it
+      // everytime we check space usage. Since bfs::space returns
+      // available amount, we use minimum available amount as threshold. 
+      boost::system::error_code ec;
+      auto info = bfs::space(path_name, ec);
+      EOS_ASSERT(!ec, chain::plugin_config_exception,
+         "Unable to get space info for ${path_name}: [code: ${ec}] ${message}",
+         ("path_name", path_name.string())
+         ("ec", ec.value())
+         ("message", ec.message()));
+
+      auto min_available_bytes = (100 - used_space_threshold_in_percentage) * (info.capacity / 100); // (100 - used_space_threshold_in_percentage) is the percentage of minimum number of available bytes the file system must maintain  
+
+      // Add to the list
+      filesystem_info fs = { statbuf.st_dev, min_available_bytes, path_name };
+      filesystems.push_back(fs);
```

Comment:
I would use ```filesystems.emplace_back(statbuf.st_dev, min_available_bytes, path_name);``` this would ensure that no copies are made.

Reply:
Thanks Bucky for your comments! I will incorporate them together with restructuring to address Bart's comments.